### PR TITLE
[IMP] web,*: rework list view column widths logic

### PR DESCRIPTION
**Pull request by aab-odoo**
Before this commit, column widths were sometimes not ideal. In
particular, there was sometimes a big difference in column widths
between a table with records, and the same table without records.
This was because we didn't encode maximum widths (e.g. a date
column should never be more than a given width, because we know
how much space is required to display a date).

This commit takes the shot to rework the column widths logic.
Columns now have a min width and optionally a max width. Except if
we don't have a choice, the final width will always be within that
range. It may happen that we don't comply with the max width, if
there're not enough fields to fill the 100% width of the table (and
if all fields have a max width, obviously).

We also tweaked the min and max widths of several field types. For
instance, boolean, integer and floats are allowed a bit more space
now, which helps with displaying their label (in the thead).

The column widths logic has been extracted in a hook, s.t. the
whole logic is encapsulated in a single file, dedicated to it. We
did the same for its tests.

Task 3866837

### CODE REVIEW

File: addons/stock/static/src/views/list/auto_column_width_list_renderer.js [hunk 1]
```diff
@@ -1,18 +1,8 @@
 /** @odoo-module **/
 
 import { ListRenderer } from "@web/views/list/list_renderer";
-import { useEffect } from "@odoo/owl";
 
 export class AutoColumnWidthListRenderer extends ListRenderer {
     static props = [...ListRenderer.props];
-
```
Comment (aab-odoo) [inline addons/stock/static/src/views/list/auto_column_width_list_renderer.js:8]: ask cth to check

File: addons/web/static/src/views/list/list_renderer.xml [hunk 2]
```diff
@@ -24,6 +24,7 @@
                                 t-on-pointerup="onColumnTitleMouseUp"
                                 t-on-click="() => this.onClickSortColumn(column)"
                                 t-on-keydown="(ev) => this.onCellKeydown(ev)"
+                                t-att-data-tooltip="!isDebugMode ? column.label : false"
                                 t-att-data-tooltip-template="isDebugMode ? 'web.FieldTooltip' : false"
```
Comment (aab-odoo) [inline addons/web/static/src/views/list/list_renderer.xml:28]: add label in debug

File: addons/web/static/src/views/list/column_width_hook.js [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+            if (_columnWidths[thIndex] > minWidth && canShrink) {
+                shrinkableColumns.push({ thIndex, minWidth });
+                totalAvailableSpace += _columnWidths[thIndex] - minWidth;
+            }
+        }
+        if (diff > totalAvailableSpace) {
+            // We can't find enough space => set all columns to their min width, and there'll be an
+            // horizontal scrollbar
+            for (const { thIndex, minWidth } of shrinkableColumns) {
+                _columnWidths[thIndex] = minWidth;
```
Comment (aab-odoo) [inline addons/web/static/src/views/list/column_width_hook.js:146]: why reduce if there's a scrollbar?

File: addons/web/static/src/views/list/column_width_hook.js [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+        if (diff >= 1) {
+            // All columns have a maxWidth and have been expanded to their max => expand them more
+            for (let columnIndex = 0; columnIndex < columns.length; columnIndex++) {
+                const thIndex = columnIndex + columnOffset;
+                _columnWidths[thIndex] += diff / columns.length;
+            }
+        }
+    }
+    return _columnWidths;
+}
```
Comment (aab-odoo) [inline addons/web/static/src/views/list/column_width_hook.js:202]: bug sale order to check

**Comment by aab-odoo** [comment]
robodoo rebase-merge

File: addons/hr_skills/static/src/fields/resume_one2many/resume_one2many.js [hunk 5]
```diff
@@ -12,6 +12,7 @@ export class ResumeListRenderer extends CommonSkillsListRenderer {
     static template = "hr_skills.ResumeListRenderer";
     static rowsTemplate = "hr_skills.ResumeListRenderer.Rows";
     static recordRowTemplate = "hr_skills.ResumeListRenderer.RecordRow";
+    static useMagicColumnWidths = false;
```
Comment (aab-odoo) [inline addons/hr_skills/static/src/fields/resume_one2many/resume_one2many.js:15]: to check

File: addons/hr_skills/static/src/fields/skills_one2many/skills_one2many.js [hunk 6]
```diff
@@ -58,17 +58,6 @@ export class SkillsListRenderer extends CommonSkillsListRenderer {
         });
     }
 
-    calculateColumnWidth(column) {
```
Comment (aab-odoo) [inline addons/hr_skills/static/src/fields/skills_one2many/skills_one2many.js:61]: to check

**Comment by Polymorphe57** [comment]
@robodoo r+

**Pull request closed without merging**